### PR TITLE
parametrized port autoneg tests by used port speeds

### PR DESCRIPTION
**Pull request by antonptashnik**
### Description of PR

Summary: parametrized port autoneg tests by used port speeds
Fixes # (issue)
- parametrized port autoneg tests by used port speeds

Port autoneg test check port autoneg feature by setting different speeds in some order on both ends, SONiC DUT and fanout. The speeds are being set in a loop so actual speeds tests use are not visible to a test report. To resolve it the tests were parametrized by interface/speed so now a test report is clear about what interface/speed used for a given test. Example output fragment:
```
test_auto_negotiation_fanout_advertises_each_speed[dut|Ethernet8|40000-ALL_SPEEDS_BY_ALL_LITERAL]
test_auto_negotiation_fanout_advertises_each_speed[dut|Ethernet8|40000-ALL_SPEEDS_BY_SPEEDS_LIST]
test_auto_negotiation_fanout_advertises_each_speed[dut|Ethernet8|100000-ALL_SPEEDS_BY_ALL_LITERAL]
test_force_speed[dut|Ethernet8|40000]
```
- skipped test runs for port speeds limited by a connected SFP
- handled speed change prompt for EOS fanouts
- made test_force_speed to set port FEC explicitly after setting port speed
Port may stay with an FEC unsupported for a given speed since speed/FEC are set separately. To avoid such case a FEC value is set explicitly after speed setup
- fixed https://github.com/sonic-net/sonic-mgmt/issues/5447

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012

#### What is the motivation for this PR?
Make a test report more informative; skip extra test runs
#### How did you do it?
Parametrized tests by port speeds
#### How did you verify/test it?
py.test --inventory=../ansible/lab,../ansible/veos --testbed_file=../ansible/testbed.csv --module-path=../ansible/library -v -rA --topology=t1,any platform_tests/test_auto_negotiation.py

### CODE REVIEW

**Comment by SavchukRomanLv** [comment]
@prgeor, @vdahiya12 could you please review? Thank you in advance!

**Comment by SavchukRomanLv** [comment]
@prgeor, @vdahiya12 could you please review?

**Comment by antonptashnik** [comment]
@sujinmkang @prgeor take a look please

**Comment by prgeor** [comment]
@keboliu @Junchao-Mellanox can you please review?

**Comment by antonptashnik** [comment]
@ZhaohuiS @wangxin please review and suggest any changes if needed. Thanks in advance!

**Comment by antonptashnik** [comment]
@ZhaohuiS @wangxin please review

File: tests/platform_tests/test_auto_negotiation.py [hunk 1]
```diff
@@ -186,239 +140,212 @@ def check_ports_up(duthost, dut_ports, expect_speed=None):
...
-    highest_speed = supported_speeds[-1]
-    actual_speed = int_status[dut_port]['speed'][:-1] + '000'
+    common_supported_speeds = enum_dut_portname_module_fixture['speeds']
+    highest_speed = max(map(lambda p: int(p), common_supported_speeds))
+    actual_speed = int(int_status[dut_port]['speed'][:-1] + '000')
     pytest_assert(actual_speed == highest_speed, 'Actual speed is not the highest speed')
 
-def test_auto_negotiation_dut_advertises_each_speed(enum_dut_portname_module_fixture):
+
+def is_sfp_speed_supported(duthost, if_name, port_speed):
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/test_auto_negotiation.py:198]: move to port_utils.py?

File: tests/platform_tests/test_auto_negotiation.py [hunk 2]
```diff
@@ -186,239 +140,212 @@ def check_ports_up(duthost, dut_ports, expect_speed=None):
...
+    sfp_type = re.search(r'\w?SFP[\d\w+]{0,3}', out["stdout_lines"][0]).group()
+    if not sfp_type:
+        return True
+    out = duthost.shell('redis-cli -n 4 HGET "PORT|{}" "lanes"'.format(if_name))
+    n_lanes = len(out["stdout_lines"][0].split(','))
+    per_lane_speed = int(port_speed) // n_lanes
+    return per_lane_speed <= 25000 or sfp_type in pam4_supporting_sfps
+
+
+def skip_speed_if_unsupported(duthost, if_name, port_speed):
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/test_auto_negotiation.py:215]: do we really need this function? Can we just call pytest_require(is_sfp_speed_supported?)

File: tests/platform_tests/test_auto_negotiation.py [hunk 3]
```diff
@@ -186,239 +140,212 @@ def check_ports_up(duthost, dut_ports, expect_speed=None):
...
-        if pos == -1:
-            return None
-        speeds_str = output[pos+1:-1]
-        speeds = list(set([speed.split('G')[0] + '000' for speed in speeds_str.split(',')]))
-        cls.supported_speeds[(duthost, dut_port_name)] = speeds
-        return speeds
+    duthost.shell('config interface speed {} {}'.format(dut_port, speed))
+    logger.info('Wait until the port status is up, expected speed: {}'.format(speed))
+    
+    duthost.set_port_fec(dut_port, fec_mode)
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/test_auto_negotiation.py:333]: where is set_port_fec defined?

File: tests/platform_tests/test_auto_negotiation.py [hunk 4]
```diff
@@ -186,239 +140,212 @@ def check_ports_up(duthost, dut_ports, expect_speed=None):
...
+        speed)
+
+    pytest_assert(
+        wait_result, '{} are still down. Advertised speeds: DUT = {}, fanout = {}'
+        .format(dut_port, dut_advertised_speeds, speed))
+    fanout_actual_speed = fanout.get_speed(fanout_port)
+    pytest_assert(fanout_actual_speed == speed, 'expected fanout speed: {}, but got {}'.format(speed, fanout_actual_speed))
+
+
+FEC_FOR_SPEED = {
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/test_auto_negotiation.py:299]: move to top?

File: tests/platform_tests/test_auto_negotiation.py [hunk 5]
```diff
@@ -186,239 +140,212 @@ def check_ports_up(duthost, dut_ports, expect_speed=None):
...
+    25000: 'fc',
+    50000: 'fc',
+    100000: 'rs',
+    200000: 'rs',
+    400000: 'rs'
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/test_auto_negotiation.py:304]: why do we need these hard coded FEC for different speed?

Comment (antonptashnik) [reply tests/platform_tests/test_auto_negotiation.py]: peers might try to auto-set different FECs for a given speed which will cause force_speed test to fail. So I decided to set both speed and FEC explicitly

File: tests/platform_tests/test_auto_negotiation.py (hunk 3 above)
Comment (antonptashnik) [reply tests/platform_tests/test_auto_negotiation.py:313]: added a missing code

**Comment by antonptashnik** [comment]
@Junchao-Mellanox re-review please. I took into account your comments

**Comment by Junchao-Mellanox** [comment]
LGTM. Hi @roysr-nv , could you please also review?

**Review by roysr-nv — COMMENTED** [review]
The enhancements of the existing test cases is useless without addressing the following issue:
https://github.com/sonic-net/sonic-mgmt/issues/5447

**Comment by antonptashnik** [comment]
> The enhancements of the existing test cases is useless without addressing the following issue: #5447

this PR fixes the issues mentioned in https://github.com/sonic-net/sonic-mgmt/issues/5447

**Comment by roysr-nv** [comment]
> > The enhancements of the existing test cases is useless without addressing the following issue: #5447
> 
> this PR fixes the issues mentioned in #5447

Now it's clear, since it's mentioned in the PR description- thanks!

**Comment by antonptashnik** [comment]
@ZhaohuiS please review and notify in case any changes needed. Thanks in advance!

**Comment by antonptashnik** [comment]
@ZhaohuiS @roysr-nv please review and merge if no changes needed. Thanks !

File: tests/common/devices/sonic.py [hunk 6]
```diff
@@ -2137,6 +2137,25 @@ def get_port_channel_status(self, port_channel_name):
...
+        return out["stdout_lines"][0]
+
+    def set_port_fec(self, portname, state):
+        if not state:
+            state = 'none'
+        self.shell('sudo config interface fec {} {}'.format(portname, state))
+
+    def count_portlanes(self, portname):
+        out = self.shell('redis-cli -n 4 HGET "PORT|{}" "lanes"'.format(portname))
+        lanes = out["stdout_lines"][0].split(',')
```
Comment (ZhaohuiS) [inline tests/common/devices/sonic.py:2151]: @antonptashnik If the output is something error when running redis-cli, such as timeout, should we have exception handle here?

File: tests/conftest.py [hunk 7]
```diff
@@ -1330,8 +1330,33 @@ def pytest_generate_tests(metafunc):        # noqa E302
...
+            metafunc.parametrize(
+                "enum_dut_portname_module_fixture",
+                autoneg_tests_data,
+                scope="module",
+                ids=format_id,
+                indirect=True
+            )
+
+        if "enum_speed_per_dutport_fixture" in metafunc.fixturenames:
+            def format_id(param):
```
Comment (ZhaohuiS) [inline tests/conftest.py:1349]: @antonptashnik Why keep two same functions here? Is it better to move it out and define a common one?

**Comment by ZhaohuiS** [comment]
@antonptashnik Thank you for your hard work on the enhancement and fixing #5447, much appreciate it! I left a minor comment, could you please take a look?
BTW, I see you addressed many comments during the review, that's why this PR became much bigger and complicated since it has many different changes. It's fine for this pretty old PR.
May I ask you address one problem in a smaller PR next time if possible? Thanks!

**Comment by antonptashnik** [comment]
> @antonptashnik Thank you for your hard work on the enhancement and fixing #5447, much appreciate it! I left a minor comment, could you please take a look? BTW, I see you addressed many comments during the review, that's why this PR became much bigger and complicated since it has many different changes. It's fine for this pretty old PR. May I ask you address one problem in a smaller PR next time if possible? Thanks!

Yes sure, I'll try to keep future PRs smaller and more focused. Thanks for your review!

**Comment by antonptashnik** [comment]
@ZhaohuiS I made the changes you requested. Re-review please

**Review by ZhaohuiS — APPROVED** [review]
LGTM

**Pull request merged**